src/reactviews/pages/PublishProject/components/PublishTargetSection.tsx

```diff
@@ -3,24 +3,269 @@
+    const targetComponent = usePublishDialogSelector(
+        (s) => s.formComponents[PublishFormFields.PublishTarget],
+    );
+    const targetValue = usePublishDialogSelector(
+        (s) => s.formState[PublishFormFields.PublishTarget],
+    );
+
+    const isContainer = targetValue === PublishTarget.LocalContainer;
+
+    // Container-specific fields (only select when needed)
+    const portComponent = usePublishDialogSelector(
+        (s) => s.formComponents[PublishFormFields.ContainerPort],
+    );
+    const portValue = usePublishDialogSelector((s) => s.formState[PublishFormFields.ContainerPort]);
+
+    const passwordComponent = usePublishDialogSelector(
+        (s) => s.formComponents[PublishFormFields.ContainerAdminPassword],
+    );
+    const passwordValue = usePublishDialogSelector(
+        (s) => s.formState[PublishFormFields.ContainerAdminPassword],
+    );
+
+    const confirmPasswordComponent = usePublishDialogSelector(
+        (s) => s.formComponents[PublishFormFields.ContainerAdminPasswordConfirm],
+    );
+    const confirmPasswordValue = usePublishDialogSelector(
+        (s) => s.formState[PublishFormFields.ContainerAdminPasswordConfirm],
+    );
+
+    const imageTagComponent = usePublishDialogSelector(
+        (s) => s.formComponents[PublishFormFields.ContainerImageTag],
+    );
+    const imageTagValue = usePublishDialogSelector(
+        (s) => s.formState[PublishFormFields.ContainerImageTag],
+    );
+
+    const licenseComponent = usePublishDialogSelector(
+        (s) => s.formComponents[PublishFormFields.AcceptContainerLicense],
+    );
+    const licenseValue = usePublishDialogSelector(
+        (s) => s.formState[PublishFormFields.AcceptContainerLicense],
+    );
+
+    // Password visibility state management
+    const [showAdminPassword, setShowAdminPassword] = useState(false);
+    const [showConfirmPassword, setShowConfirmPassword] = useState(false);
+
+    // Local password state to prevent cursor jumping to end of the text
+    const [localAdminPassword, setLocalAdminPassword] = useState(passwordValue?.toString() || "");
+    const [localConfirmPassword, setLocalConfirmPassword] = useState(
+        confirmPasswordValue?.toString() || "",
+    );
+
+    // Sync local password state with external state when external values change
+    useEffect(() => {
+        setLocalAdminPassword(passwordValue?.toString() || "");
+    }, [passwordValue]);
 
     useEffect(() => {
-        setLocalValue(value);
-    }, [value]);
+        setLocalConfirmPassword(confirmPasswordValue?.toString() || "");
+    }, [confirmPasswordValue]);
 
-    if (!publishCtx) {
-        return undefined;
+    // Auto-populate defaults and revalidate passwords
+    useEffect(() => {
+        if (!publishCtx || !isContainer) {
+            return;
+        }
+
+        // Default container port if not set
+        if (!portValue) {
+            publishCtx.formAction({
```

Comment:
Please take a look at how the other form components are structured.  We shouldn't be repeating onChange and onBlur for every input.